text/0098-metric-instruments-explained.md

```diff
@@ -0,0 +1,245 @@
+
+Asynchronous delta instruments are excluded from the standard based on the lack of motivating examples, but we could also justify this as a desire to keep asynchronous callbacks stateless. An observer has to have memory in order to compute deltas; it is simpler for asynchronous code to report cumulative values.
+
+With six instruments in total, one may be curious--how does the historical Metrics API term _Gauge_ translate into this specification?  _Gauge_, in Metrics API terminology, may cover all of these instrument use-cases with the exception of `Counter`.  As defined in [OTEP 88][otep-88], the OpenTelemetry Metrics API will disambiguate these use-cases by requiring *single purpose instruments*.  The choice of instrument implies a default interpretation, a standard aggregation, and suggests how to treat Metric data in observability systems, out of the box.  Uses of `Gauge` translate into the various OpenTelemetry Metric instruments depending on what kind of values is being captured and whether the measurement is made synchronously or not.
+
+Summarizing the naming scheme:
+
+- If you've measured an amount of something that adds up to a total, where you are mainly interested in that total, use one of the additive instruments:
+  - If synchronous and monotonic, use `Counter` with non-negative values
+  - If synchronous and not monotonic, use `UpDownCounter` with arbitrary values
+  - If asynchronous and a cumulative, monotonic sum is measured, use `SumObserver`
+  - If asynchronous and a cumulative, arbitrary sum is measured, use `UpDownSumObserver`
+- If the measurements are non-additive or additive with an interest in the distribution, use an instantaneous instrument:
+  - If synchronous, use `ValueRecorder` to record a value that is part of a distribution
+  - If asynchronous use `ValueObserver` to record a single measurement nearing the end of a collection interval.
+
+### Sync vs Async instruments
+
+Synchronous instruments are called in a request context, meaning they potentially have an associated tracing context and distributed correlation values.  Multiple metric events may occur for a synchronous instrument within a given collection interval.
+
+Asynchronous instruments are reported by a callback, once per collection interval, and lack request context.  They are permitted to report only one value per distinct label set per period.  If the application observes multiple values in a single callback, for one collection interval, the last value "wins".
+
+### Temporal quality
+
+Measurements can be described in terms of their relationship with time.
```

Comment:
I found comments with examples that explain temporal qualities in https://github.com/open-telemetry/opentelemetry-proto/blob/28e2774262a651c4c2e701c3a05820c3e43ead71/opentelemetry/proto/metrics/v1/metrics.proto#L156 very helpful. It may be useful to have similar examples in this OTEP.

Reply:
We've had a lot of debate about this topic in the last few days. Surprisingly much of it, and the leading conclusion is that we should avoid these terms in the specification. Without rewriting this proposal, I've added to clarify that these terms will not be used going forward:

> Note: although this term logically applies and is used throughout this OTEP, discussion in the Metrics SIG meeting (4/30/2020) leads us to exclude this term from use in documenting the Metric API.  The explanation of terms here is consistent with the [terminology used in the protocol], but we will prefer to use these adjectives to describe properties of an aggregation, not properties of an instrument (despite this document continuing to use the terms freely).  In the API specification, this distinction will be described using "additive synchronous" in contrast with "additive asynchronous".

I have also written in the forthcoming API spec for version 0.4 this on the same topic:

> The additive instruments (`Counter`, `UpDownCounter`, `SumObserver`,
`UpDownSumObserver`) use Sum aggregation, by default.  Details about
computing a Sum aggregation vary, but from the user's perspective this
means they will be able to monitor the sum of values entered.  The
distinction between synchronous and asynchronous instruments is
crucial to specifying how exporters work, a topic that is covered in
the [SDK specification](TODO).